### PR TITLE
rpi_pico: Added flash support

### PR DESCRIPTION
**Pull request by yonsch**
Add support for flash read/write/erase on rpi_pico. Tested with the `flash_shell` sample.
Note: Writes are currently inefficient, as they are performed on a page basis. This is because `pico-sdk` does not currently provide an API for writing smaller ranges.

### CODE REVIEW

**Comment by Laczen** [comment]
@yonsch, why don't you use a write block size of 256 byte. This will limit the possibilities of using the flash but it will be correct and avoid unnecessary erases when trying to emulate 1 byte writing support. If you do wish to support 1 byte writing it should be required to read whatever is on the flash (instead of setting it to the erase value), apply the changes, and write to flash.

**Comment by Laczen** [comment]
@yonsch, is it safe to write to flash when executing from the same flash?

File: drivers/flash/flash_rpi_pico.c [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+#include <hardware/flash.h>
+
+LOG_MODULE_REGISTER(flash_rpi, CONFIG_FLASH_LOG_LEVEL);
+
+#define DT_DRV_COMPAT raspberrypi_pico_flash_controller
+
+#define PAGE_SIZE 256
+#define ERASE_VALUE 0xff
+#define FLASH_PAGE_MASK (0xffffffff - (PAGE_SIZE - 1))
+#define FLASH_SIZE (CONFIG_FLASH_SIZE * 1024)
```
Comment (ycsin) [inline drivers/flash/flash_rpi_pico.c:19]: maybe `KB(CONFIG_FLASH_SIZE)`? 
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/util.h#L447

File: drivers/flash/flash_rpi_pico.c [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+static uint8_t page_buffer[PAGE_SIZE];
+
+static const struct flash_parameters flash_rpi_parameters = {
+	.write_block_size = 1,
+	.erase_value = ERASE_VALUE,
+};
+
+static bool is_invalid_range(off_t offset, uint32_t size)
+{
+	return offset + size > FLASH_SIZE;
```
Comment (ycsin) [inline drivers/flash/flash_rpi_pico.c:30]: make the precedence explicit:
```c
(offset + size) > FLASH_SIZE;
```

File: drivers/flash/flash_rpi_pico.c [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+static int flash_rpi_read(const struct device *dev, off_t offset,
+				void *data, size_t size)
+{
+	if (is_invalid_range(offset, size)) {
+		return -EINVAL;
+	}
+
+	if (!size) {
```
Comment (ycsin) [inline drivers/flash/flash_rpi_pico.c:40]: MISRA C:2004, 13.2 - Tests of a value against zero should be made explicit, unless the operand is effectively Boolean.

File: drivers/flash/flash_rpi_pico.c [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+					ERASE_VALUE,
+					PAGE_SIZE - size - left_pad_len);
+		}
+
+		flash_range_program(offset & FLASH_PAGE_MASK, page_buffer,
+					PAGE_SIZE);
+		offset = (offset & FLASH_PAGE_MASK) + PAGE_SIZE;
+		size -= write_count;
+		_data += write_count;
+
```
Comment (ycsin) [inline drivers/flash/flash_rpi_pico.c:97]: remove

File: drivers/flash/flash_rpi_pico.c [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+		memcpy(page_buffer, _data, size);
+		memset(page_buffer + size, ERASE_VALUE, PAGE_SIZE - size);
+		flash_range_program(offset, page_buffer, PAGE_SIZE);
+	}
+
+	return 0;
+}
+
+static int flash_rpi_erase(const struct device *dev, off_t offset,
+				size_t size)
```
Comment (ycsin) [inline drivers/flash/flash_rpi_pico.c:124]: can be made into one line, the col limit is 100

File: drivers/flash/flash_rpi_pico.c [hunk 6]
```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2022 Yonatan Schachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <errno.h>
```
Comment (ycsin) [inline drivers/flash/flash_rpi_pico.c:6]: is this required explicitly? Not 100% sure but I guess it should be included in some other headers that are already included in this file

File: drivers/flash/flash_rpi_pico.c [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+		 * The B part of the data is page aligned, so it can be
+		 * written with pico-sdk API.
+		 */
+		flash_range_program(offset, _data, size & FLASH_PAGE_MASK);
+		size -= size & FLASH_PAGE_MASK;
+		offset += size & FLASH_PAGE_MASK;
+		_data += size & FLASH_PAGE_MASK;
+	}
+
+	if (size) {
```
Comment (ycsin) [inline drivers/flash/flash_rpi_pico.c:111]: might be better to test if size > 0 as well:
MISRA C:2004, 13.2 - Tests of a value against zero should be made explicit, unless the operand is effectively Boolean.

**Comment by yonsch** [comment]
@Laczen
> why don't you use a write block size of 256 byte

I wanted this to be future compatible for when this feature will be available, but it turns out that the flash works with page writes, so there is no way around a write size of 256.
>  is it safe to write to flash when executing from the same flash?

Yes, because the relevant functions are executed from RAM. Disabling interrupts was also added to prevent hard faults when ISRs are called during flash operations.

File: drivers/flash/flash_rpi_pico.c (hunk 6 above)
Comment (yonsch) [reply drivers/flash/flash_rpi_pico.c:6]: Probably not, but I prefer not to be dependent on one of those headers which might be removed in the future.

File: drivers/flash/Kconfig.rpi_pico [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2022 Yonatan Schachter
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config FLASH_RPI_PICO
+	bool "Raspberry Pi Pico flash driver"
+	default y
+	depends on FLASH && SOC_SERIES_RP2XXX
```
Comment (galak) [inline drivers/flash/Kconfig.rpi_pico:8]: `depends on DT_HAS_RASPBERRYPI_PICO_FLASH_CONTROLLER_ENABLED`  should replace the 'depends on FLASH && SOC_SERIES_RP2XXXX`

**Comment by Laczen** [comment]
> Added "real" byte-aligned writes. The low level functions were taken from the SDK and bootrom, and adapted to Zephyr (functionally and stylistically). I suppose this is fine in terms of licensing (the sdk is BSD-3 Clause), but would like someone to verify this.

@yonsch, regarding the "real" byte-aligned writes: it is strange that the write-block-size is set to the page size while the routines allow smaller writes, is this intended ? I didn't find the `_write_partial` routines, but do they work by reading, modifying and writing back ? If they do it might be a good idea to set the write-block-size to 16, this will enable support for most flash usage, but still keep the amount of rewrites reduced.

File: drivers/flash/flash_rpi_pico.c [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+
+#define DT_DRV_COMPAT raspberrypi_pico_flash_controller
+
+#define PAGE_SIZE 256
+#define SECTOR_SIZE DT_PROP(DT_CHILD(DT_NODELABEL(flash_controller), flash_0), erase_block_size)
+#define ERASE_VALUE 0xff
+#define FLASH_SIZE KB(CONFIG_FLASH_SIZE)
+
+static const struct flash_parameters flash_rpi_parameters = {
+	.write_block_size = PAGE_SIZE,
```
Comment (lgehreke) [inline drivers/flash/flash_rpi_pico.c:34]: Write block size is given as 1 in the devicetree but as 256 here?

File: drivers/flash/flash_rpi_pico.c [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+{
+	ARG_UNUSED(dev);
+
+	return &flash_rpi_parameters;
+}
+
+#if CONFIG_FLASH_PAGE_LAYOUT
+
+static const struct flash_pages_layout flash_rpi_pages_layout = {
+	.pages_count = FLASH_SIZE / PAGE_SIZE,
```
Comment (Laczen) [inline drivers/flash/flash_rpi_pico.c:275]: Shouldn't this use `SECTOR_SIZE` instead of `PAGE_SIZE` ?

File: drivers/flash/flash_rpi_pico.c [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+	.pages_size = PAGE_SIZE,
```
Comment (Laczen) [inline drivers/flash/flash_rpi_pico.c:276]: Same as above.

**Comment by Laczen** [comment]
@yonsch, nice work. Regarding my previous remark about the `write-block-size`, the device should support 1 byte writes, so this is no problem.

File: drivers/flash/flash_rpi_pico.c [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+	}
+
+	if (size == 0) {
+		return 0;
+	}
+
+	key = irq_lock();
+
+	if (offset % PAGE_SIZE > 0) {
+		bytes_to_write = PAGE_SIZE - (offset % PAGE_SIZE);
```
Comment (lgehreke) [inline drivers/flash/flash_rpi_pico.c:212]: Doesn't this break if we do not write page aligned but still not to the end of the page, e.g. with an offset of 72 to the page start but a size of 7?

File: drivers/flash/flash_rpi_pico.c [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+}
+
+void __no_inline_not_in_flash_func(flash_put_get_wrapper)(uint8_t cmd, const uint8_t *tx,
+					uint8_t *rx, size_t count) {
+	flash_cs_force(OUTOVER_LOW);
+	ssi->dr0 = cmd;
+	flash_put_get(tx, rx, count, 1);
+}
+
+void __no_inline_not_in_flash_func(flash_put_cmd_addr)(uint8_t cmd, uint32_t addr) {
```
Comment (lgehreke) [inline drivers/flash/flash_rpi_pico.c:130]: This function is [inlined](https://github.com/raspberrypi/pico-bootrom/blob/master/bootrom/program_flash_generic.c#L160) for timing reasons. Why the __noinline here?

File: drivers/flash/flash_rpi_pico.c (hunk 10 above)
Comment (yonsch) [reply drivers/flash/flash_rpi_pico.c:275]: > Here, "page" means the smallest erasable area on the flash device.
So you're right

File: drivers/flash/CMakeLists.txt [hunk 14]
```diff
@@ -28,6 +28,16 @@ zephyr_library_sources_ifdef(CONFIG_FLASH_MCUX_FLEXSPI_HYPERFLASH flash_mcux_fle
...
 zephyr_library_sources_ifdef(CONFIG_SOC_FLASH_SMARTBOND flash_smartbond.c)
 zephyr_library_sources_ifdef(CONFIG_FLASH_CAD_QSPI_NOR flash_cadence_qspi_nor.c flash_cadence_qspi_nor_ll.c)
+zephyr_library_sources_ifdef(CONFIG_FLASH_RPI_PICO flash_rpi_pico.c)
+
+if(CONFIG_FLASH_RPI_PICO)
+  # Suppress pointer arithmetic warning caused by a function pointer trick
+  set_source_files_properties(
+    flash_rpi_pico.c
+    PROPERTIES
+    COMPILE_FLAGS $<TARGET_PROPERTY:compiler,warning_no_pointer_arithmetic>
```
[the remark this answers is not in the document]
Comment (yonsch) [reply drivers/flash/CMakeLists.txt]: I cast the pointer to a uint32_t for the addition, then cast to a function pointer. Thanks.

**Review by lgehreke — DISMISSED** [review]
Thanks for the driver. Tested it with settings API, Flash shell and mcuboot.

File: drivers/flash/flash_rpi_pico.c [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+static ssi_hw_t *const ssi = (ssi_hw_t *)XIP_SSI_BASE;
+static uint32_t boot2_copyout[BOOT2_SIZE_WORDS];
+static bool boot2_copyout_valid;
+
+static void __no_inline_not_in_flash_func(flash_init_boot2_copyout)(void)
+{
+	if (boot2_copyout_valid)
+		return;
+	for (int i = 0; i < BOOT2_SIZE_WORDS; ++i)
+		boot2_copyout[i] = ((uint32_t *)XIP_BASE)[i];
```
Comment (soburi) [inline drivers/flash/flash_rpi_pico.c:66]: The register address should define in dts.

**Comment by yonsch** [comment]
@soburi Thank you for your review. I replaced `XIP_BASE` with `FLASH_BASE`, which is based on devicetree. However, I did not touch `XIP_SSI_BASE`, since I'm not sure it should be exposed to the devicetree. The user shouldn't really play with the SSI, and it's only used through this driver (this particular snippet is from the HAL, so it should be okay). I don't think we should create a compatible for the SSI just for this instance.

**Comment by carlescufi** [comment]
@soburi please approve if happy

File: drivers/flash/flash_rpi_pico.c [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+	if (!is_valid_range(offset, size)) {
+		LOG_ERR("Write range exceeds the flash boundaries. Offset=%#lx, Size=%u",
+				offset, size);
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	if (offset % PAGE_SIZE > 0) {
+		bytes_to_write = MIN(PAGE_SIZE - (offset % PAGE_SIZE), size);
```
Comment (de-nordic) [inline drivers/flash/flash_rpi_pico.c:223]: Assuming the PAGE_SIZE is power of 2, you can use binary operators here `PAGE_SIZE - (offset & (PAGE_SIZE -1))`, you did similar thing in line 231

File: drivers/flash/flash_rpi_pico.c [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+#include <hardware/structs/xip_ctrl.h>
+#include <hardware/resets.h>
+#include <pico/bootrom.h>
+
+LOG_MODULE_REGISTER(flash_rpi_pico, CONFIG_FLASH_LOG_LEVEL);
+
+#define DT_DRV_COMPAT raspberrypi_pico_flash_controller
+
+#define PAGE_SIZE 256
+#define SECTOR_SIZE DT_PROP(DT_CHILD(DT_NODELABEL(flash_controller), flash_0), erase_block_size)
```
Comment (de-nordic) [inline drivers/flash/flash_rpi_pico.c:30]: I think those should be named: `PAGE_WRITE_SIZE` and `SECTOR_ERASE_SIZE`.
The problem I have with just `PAGE_SIZE` and `SECTOR_SIZE` is that there is a real naming mess in non volatile memory storage subsystem.
Sometimes page means write block, sometimes erase size. In other instances sector is uses as write/erase or as smaller erase unit, where page consists of sectors.
The `PAGE_WRITE_SIZE` and `SECTOR_ERASE_SIZE` ain't perfect but at least it tells what the thing is used for, so you get less confused when jumping between drivers.

**Review by de-nordic — DISMISSED** [review]
Look OK. Thank you for the contribution.

Note on the `PAGE_WRITE_SIZE` and `SECTOR_ERASE_SIZE`: although I  prefer the verb WRITE/ERASE within the name of the identifier, to clarify what the identifier is supposed to do rather than magic interpretation of sector as erase unit and page as write unit, I understand that this is this is how the device specification names them.  If it will be desired by you or other reviewers to revert the names to original, I will approve the code anyway.

**Comment by Laczen** [comment]
@yonsch, the usage of `PAGE_WRITE_SIZE` and `SECTOR_ERASE_SIZE` introduce yet another terminology that only adds to the confusion. If possible I would like you to revert this to `PAGE_SIZE` and `SECTOR_SIZE` that can be easily found in the device specifications.

**Review by Laczen — APPROVED** [review]
Thanks

**Review by lgehreke — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Review by soburi — APPROVED** [review]

**Pull request merged**